### PR TITLE
AssemblyScript Implementation

### PR DESCRIPTION
**Pull request by donmahallem**
## Description

This version is an adaption in AssemblyScript

## Contributing requirements

* [x] I read the contribution guidelines in CONTRIBUTING.md.
* [x] I placed my solution in the correct category folder.
* [x] I added a README.md or Dockerfile.
* [x] I selected `drag-race` as the target branch.
* [x] All code herein is licensed compatible with BSD-3.

### CODE REVIEW

**Comment by donmahallem** [comment]
Currently I am facing an issue might someone else can fix for me:

If I do use the docker version it runs roughly half as fast as the version without docker

**Comment by marghidanu** [comment]
Here's the benchmark on an M1 MacBook Air:

```
Passes: 10620, Time: 10.0, Avg: 0.0009416195680387318, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: true
```

Ans using Docker on the same machine I get these results:

```
Passes: 8251, Time: 10.0, Avg: 0.0012119743041694165, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: true
```

File: PrimeAssemblyScript/package.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "primeassemblyscript",
+  "version": "1.0.0",
+  "description": "Prime Sieve Implementation in AssemblyScript",
+  "main": "index.js",
+  "scripts": {
+    "build": "asc -O3",
+    "build:bench": "asc -O3 --shrinkLevel 0 --noAssert --converge",
```
Comment (MaxGraey) [inline PrimeAssemblyScript/package.json]:
```suggestion
    "build:bench": "asc -O3 --noAssert --converge",
```

File: PrimeAssemblyScript/package.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    "build:debug": "asc -O3 --debug",
```
Comment (MaxGraey) [inline PrimeAssemblyScript/package.json]:
```suggestion
    "build:debug": "asc --debug",
```

File: PrimeAssemblyScript/src/sieve.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            ", Time: " + duration.toString() +
+            ", Avg: " + (duration / (passes as f32)).toString() +
+            ", Limit: " + this.sieveSize.toString() +
+            ", Count1: " + count.toString() +
+            ", Count2: " + this.countPrimes().toString() +
+            ", Valid: " + this.validateResults().toString());
+    }
+
+    countPrimes(): i32 {
+        let count: i32 = (this.sieveSize >= 2);;
```
Comment (MaxGraey) [inline PrimeAssemblyScript/src/sieve.ts]:
```suggestion
        let count = i32(this.sieveSize >= 2);
```

File: PrimeAssemblyScript/src/sieve.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        for (let i: i32 = 3; i < this.sieveSize; i += 2)
```
Comment (MaxGraey) [inline PrimeAssemblyScript/src/sieve.ts]:
```suggestion
        for (let i = 3, len = this.sieveSize; i < len; i += 2)
```

File: PrimeAssemblyScript/src/sieve.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    validateResults(): bool {
+        if (PrimeSieve.primeCounts[this.sieveSize])
+            return PrimeSieve.primeCounts[this.sieveSize] == this.countPrimes();
+        return false;
+    }
+
+    runSieve(): void {
+        let factor: i32 = 3;
+        let q: i32 = Math.sqrt(this.sieveSize) as i32;
```
Comment (MaxGraey) [inline PrimeAssemblyScript/src/sieve.ts]:
```suggestion
        let factor = 3;
        let q = Math.sqrt(this.sieveSize) as i32;
```

File: PrimeAssemblyScript/src/sieve.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+        while (factor <= q) {
+            for (let num: i32 = factor; num < this.sieveSize; num += 2) {
```
Comment (MaxGraey) [inline PrimeAssemblyScript/src/sieve.ts]:
```suggestion
            let size = this.sieveSize;
            for (let num = factor; num < size; num += 2) {
```

File: PrimeAssemblyScript/src/sieve.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                if (!this.Bits[num]) {
+                    factor = num;
+                    break;
+                }
+            }
+            for (let num: i32 = factor * factor; num < this.sieveSize; num += factor * 2)
```
Comment (MaxGraey) [inline PrimeAssemblyScript/src/sieve.ts]:
```suggestion
            for (let num = factor * factor; num < size; num += factor * 2)
```
depends on previous suggestion

**Comment by marghidanu** [comment]
Do the numbers look better now?

**Comment by donmahallem** [comment]
@MaxGraey is it a technical or styling reason that you removed the explicit types from the variables?
Besides that I agree with your suggestions.

**Comment by marghidanu** [comment]
The only thing that's missing here is having the solution in the right directory, otherwise, I think this is ready for merge

**Comment by donmahallem** [comment]
@marghidanu like this?

**Comment by marghidanu** [comment]
LGTM

**Comment by MaxGraey** [comment]
> is it a technical or styling reason that you removed the explicit types from the variables?

Yes, just suggested less verbose style

**Pull request merged**